### DISCUSSION TITLE
Break down Exiftool into 2+ classes, a raw Exiftool, and helper classes

### DISCUSSION THREAD

**Discussion by sylikc** [sylikc/pyexiftool#10] [Ideas]
I'm planning to roll back some of the PR that I merged really early on from this project.  (specifically the commits around 9c15d425)

I'm going to re-design a little bit, and put all "Helper" functions which wrap around raw Exiftool in another class.  This should help clean up the code and make it easier to extend later

The original Exiftool class was more of a "raw" class where it had all the Exiftool low level functionality.  The addition of methods that made it easier to use into the class will make it harder to maintain over time.  
* set_keywords_batch
* set_tags_batch
* get_tag_wrapper
... etc

I believe these should be in a separate class, which either uses Exiftool directly, or extends it.  It should make it easier to fork and add functionality, without making the Exiftool class unmanageably big with a bunch of helper functions

In fact, down the line, should be able to make even more streamlined functionality available... say, if you wanted to "rotate a picture by 90 degrees" right now, you have to raw pass into Exiftool class all the parameters to make it happen.  Extending a helper class could make it doable in a few lines

What do you all think?

**Comment by csparker247** [comment]
If I understand correctly, keep a low-level ExifTool class which passes raw parameters to the process, and package up the interesting but "lots of boiler plate arguments" features into a higher-level API that makes interacting a bit friendlier. I'm on board with this. 👍

**Reply by sylikc** [reply]
### Train of Thought

Well, what got me thinking was... 

When I started this fork, at commit 3db3764895e687d75b42d3ae4e554ca8664a7f6f, I was somewhat new to Python and I didn't know many people used this project.  I needed the Windows bug fixes, and in an effort to learn and be helpful to the community, decided to merge all the PRs as they were.

Turns out, it sounds like from @smarnach in https://github.com/smarnach/pyexiftool/pull/31#issuecomment-569238073 , he didn't intend to have this to be a long term project either.  It all started with being a helpful citizen on a StackOverflow question years ago https://stackoverflow.com/questions/10075115/call-exiftool-from-a-python-script

### Plan

But now I want to push the project forward, to make it more maintainable library, and one that might be more extensible (I've had more experience with packages since picking this up):
- I'm planning to essentially "roll back" PyExifTool's main class to the "raw Exiftool" class in the last upstream commit 3db3764895e687d75b42d3ae4e554ca8664a7f6f . 
- I'll break out all that "new/unnecessary for raw processing" functionality out to class/es
- This would probably become the 0.5.x line, becoming **script breaking changes for everyone who's been using this forked source**.
- The **[PyExifTool PyPI](https://pypi.org/project/PyExifTool/) users won't be affected** except for the latest and greatest version that I uploaded yesterday which includes the new 0.4.x functionality (since it's been the old upstream source for ages)
- I'll probably tag the last release (probably same as the Python 2.x) before the refactor, and put **active development on a branch** while waiting for the code to stabilize
- I might write documentation for a 'migration' plan as to where the old functionality went (I myself have code that depends on those 'extra' methods)
- Hopefully if I can stabilize the API, then can make a 1.0 release and keep Exiftool main class stable.

### Extra Motivation

I was looking at some other projects which use Exiftool as the underlying utility, say like https://github.com/EdLeafe/pyexif/blob/main/pyexif/pyexif.py , and there's lots of boiler plate code that would make common operations less error prone.

Rotation, orientation, keywords, **working with date/time values and converting them to/from native Python types** automatically... 

Those would be things I'd want to put into helper class/es

**Reply by csparker247** [reply]
That all works for me. It could be weird to have an API breaking change as a minor version, but only if you have a significant  numbers of users of your fork who rely on your features. Looking at my usage, it's all `get_batch_metadata` for reading tags and `execute` calls for writing tags, so I don't think your plans will cause any breaking changes for me.

I haven't given it much thought, but I would be interested in seeing the interface for writing tags improve. I provide a GUI where users can enter custom values for a pre-selected list of tags, all of which are optional, so there's a decent bit of boilerplate where I have to construct the ExifTool args from scratch. Maybe this is a feature I can give back to the project when you're ready for it.

**Reply by sylikc** [reply]
is your project with the GUI open source?  

Yeah I want to see how others are using it and then try to incorporate easier to use calls.  My project also had to have like the boilerplate stuff in variables, and then just send extra params... long story short, that's what'll be improved one

The `get_batch_metadata` was an added call, that will get sent to a different class.  The `execute` will be in the base class with the direct call into exiftool

**Reply by csparker247** [reply]
It's not unfortunately, but an open source release is on the road map. In the mean time, I think I can extract the relevant parts into an indicative example.

**Reply by csparker247** [reply]
Per my last post, here's my relevant code. Background is that I have two programs. One is control software for a multi-camera rig which simultaneously captures from some number of cameras. These images get written, and then I use ExifTool to add common information (rig make-model, software version, user-defined tags, etc.) and image specific metadata (camera make-model, focal length, etc.). Because a capture session can run for more than one instantaneous capture, I keep an ExifTool object around that I have to manually start and cleanup. That code looks like this:
```python
### During construction of the capture manager ###
# setup common exif tags
self._default_exif_flags = ['-overwrite_original',
                            f'-Software={self._metadata["software"]}',
                            f'-Format={self.image_format}']
# This is just a list of key=value pairs, where key is predefined to be an ExifTool flag. Value is user-defined.
for tag in self.info.exif_tags():
    if len(tag[1]) > 0:
        self._default_exif_flags.append(f'-{tag[0]}={tag[1]}')

### At capture time, for each image ###
# Add image specific tags
exif_flags = self._default_exif_flags.copy()
capture_time = dt.now(datetime.timezone.utc).strftime("%Y:%m:%d %H:%M:%SZ")
exif_flags.append(f'-xmp:dateTimeOriginal={capture_time}')
if cam.make:
    exif_flags.append(f'-Make={cam.make}')
if cam.model:
    exif_flags.append(f'-Model={cam.model}')
if cam.focal_length > 0.0:
    exif_flags.append(f'-FocalLength={cam.focal_length}')
if cam.f_num > 0.0:
    exif_flags.append(f'-FNumber={cam.f_num}')
    
# File path is always last
exif_flags.append(str(file_path))

# Encode the flags to bytes
exif_flags = [flag.encode() for flag in exif_flags]

# Start exiftool if it's not running
if not self._exif_tool.running:
    self._exif_tool.start()

# Run the command
self._exif_tool.execute(*exif_flags)
```

My second program consumes the files generated by the capture program. It's much more straightforward: glob all of the files in the capture directory and batch read their metadata into a dict:
```python
# Reading metadata
images = list(dir.glob(f'{prefix}*.{ext}'))
images.sort()
files = [str(i) for i in images]
if len(files) == 0:
    print('Error: Provided scan metadata specifies file pattern, but no files match.')
    raise RuntimeError()

with exiftool.ExifTool() as et:
    img_metadata = et.get_metadata_batch(files)
```

Hope that helps.

**Reply by csparker247** [reply]
It occurs to me that you don't have to roll back to that commit. You could just remove that functionality from the raw class as you implement it in the helper class/classes.

**Reply by sylikc** [reply]
Thanks for the relevant code.  I'll look at it a bit later.  

The problem with some of those usage like `-Make=` is that exiftool gives you like ten different ways of doing things... the usage like that changes all `Make` in every section, or can target a specific section or ... hmm, I'll just have to think about how generic is _generic_ lol.  I'll come up with something for review...

I'm not planning to do a `git revert` on that commit.  I've started a branch where I'm starting to make incompatible changes slowly.  `v0.5.x-py3-refactor` is the branch I'm on to make all these incompatible changes.

Still debating whether i want to break things that I feel could be improved... like `start` should be related to `stop` ... but right now it's named `terminate` to stop the process.

There's other little things like that.

**Reply by csparker247** [reply]
> The problem with some of those usage like `-Make=` is that exiftool gives you like ten different ways of doing things... the usage like that changes all `Make` in every section, or can target a specific section or ... hmm, I'll just have to think about how generic is _generic_ lol. I'll come up with something for review...

Yeah, that's a tough problem. I have lots of thoughts and theories, but it's all academic at this point, so I'll resist 😛. I'm happy to test whatever you come up with!

> I'm not planning to do a `git revert` on that commit. I've started a branch where I'm starting to make incompatible changes slowly. `v0.5.x-py3-refactor` is the branch I'm on to make all these incompatible changes.

Ahh, I see. I misunderstood. Good deal! 

> Still debating whether i want to break things that I feel could be improved... like `start` should be related to `stop` ... but right now it's named `terminate` to stop the process.

My natural inclination would be to follow the language of the `subprocess` API for the low-level class. So `terminate` is fine, but `run` might make more sense than `start`.

I'm pretty excited about how things are moving here!

**Comment by sylikc** [comment]
ok @csparker247 , I took your suggestion into account and renamed start() to run().  I like that idea, match what `subprocess` API already does.

As I've been stepping through different test cases, I've found that when no files are processed, the JSON decode blows up... since it can't decode an empty string.  I'm trying to decide... what to return from `execute_json`

My idea is... 
* if it's just "no files found" return `[]`
* If it's an error, return `None` or `raise` an error?

What do you think?

As I went about it, I realized that the original code redirected `stderr` to `devnull`, making troubleshooting what actually got returned from exiftool a bit haphazard.  I made lots of changes to read from stderr, but needs some testing heh

I'm going to see if I can come up with test cases to look at stderr output... and also test my code on Linux

**Reply by csparker247** [reply]
If the output is completely empty (no files processed) or some hard error occur, I'd `raise` an exception. Either the user provided a malformed command (bad paths, bad flags, etc.) or something bad happened in `exiftool`. Either way, just returning `None` gives very little debug information about what went wrong.

I think it's trickier when at least one file *does* get processed. The returned output is valid json, but it's missing the entries for the failed files and there's no placeholders (i.e. empty dicts) for the missed files. For that case, the output becomes harder to use, since there's no way of knowing which files didn't get processed without iterating the list. In this case, my inclination would be to raise an exception by default, but maybe provide an argument that lets the user override that behavior and still get the output. This would be similar to the [`check` argument for `subprocess.run`](https://docs.python.org/3/library/subprocess.html#subprocess.run).

Not sure how familiar you are with Github actions, but testing across multiple Python versions on Ubuntu is [fairly easy](https://docs.github.com/en/actions/guides/building-and-testing-python).

**Reply by sylikc** [reply]
Hmm, I'm not familiar with Actions at all, I saw that advertised a few weeks ago, will look into it next week.  Thanks for the link!

Yeah, the `execute_json_wrapper` stuff from the original PR 11 was dealing with that uncertainty of output.  Though after reviewing the code, I think the implementation leaves a lot to be desired (I'm likely to remove it down the line).  Then there's the thing about passing "*" instead of actually passing a file list... even the exiftool process stderr only tells us how many were processed and how many were skipped, but doesn't list out the actual file names...

Alright, anyways... I'll think about it.  Thanks for your input!

**Reply by csparker247** [reply]
Hmm, on my build, ExifTool prints a message to stderr indicating the file as it encounters a non-existent file, then summarizes the statistics at the end as you describe. On the command line, the messages are interspersed with the json output, so I missed it at first.

**Reply by sylikc** [reply]
alright, I figured out how to do read stderr now without accidentally blocking on windows.  I found the exiftool option to do a `-text4` so it always writes to STDERR, and has synchronization to make sure I completely empty the pipe before continuing.  I'm going to have to write a bunch of error test cases later to make sure I got all the bases covered heh

Btw, I'm renaming the `get_metadata_batch` function to `get_metadata` and removing the code for the solo version.  I found a bug where if you pass a wildcard to `get_metadata` like `*.jpg`, it still returns only 1 result, which is erroneous... so `get_metadata` will now always return a list, regardless of whether a string was given (one file) or a list was given (multi files) ... combining the functionality into one function instead of two

**Comment by ickc** [comment]
It is often good to streamline to make it more maintainable, extensible, and usable. As long as you communicate it clearly (including semantic versioning, documentation) and provide a path for people to easily adapt to the new interface, it should be good.

**Reply by sylikc** [reply]
Yeah, I'm going to have to work hard on documentation when this is all done.  I haven't done a CHANGELOG yet as I do commits to the v0.5 branch because, well, the interfaces aren't finalized yet.  (I'm trying to write tests that get full code coverage though to cover all the new functionality in tests so I don't inadvertently break past functionality)

Updating all the doc strings will be a pain :/

### Current status
* I've taken to making `ExifTool` the base class, and right now a class named `ExifToolHelper` which will encapsulate the functionality that `ExifTool` class used to do, but in a more extensible way
* Lots of the interfaces have changed as I feel like the `*_batch` functions are largely unnecessary and assumes no wildcards get passed into `filenames`.  When a wildcard does, all these functions pretty much break down and become useless
* I did things like merging code paths, `get_metadata` is essentially `get_tags` and passing in [] for tags.  This should simplify code maintenance
* I plan to maybe write a sample that extends `ExifToolHelper` to add more specific functionality (maybe, or maybe just let the programmer figure it out).

### More thinking
Right now I'm thinking about the `params=` wrappers i added way back when merging the [old PR 11](https://github.com/smarnach/pyexiftool/pull/11), and making them final.  It seems the latest PR #11 (coincidental) also wanted to add `params=`, so I might make it permanent depending on how the thing ends up panning out.

**Reply by ickc** [reply]
That said (making it easy for people to port over the new interface), from my limited experience, it is more important that the interface is useful and make sense to you. It is nice to have other people using your package, but since this is a small open source project, the most important user is the maintainer (you). As long as you find it useful and enjoy it, that keeps it alive.

For me, I care about the ability of batch processing a ton of files at once. The primary reason I use the original library is that it uses the native batch mode of exiftool in Perl (kind of opening a server and keep sending it commands.) I’m not sure it is related to the things you’re modifying, but I can keep my own fork for that purpose.

P.S. By my own test, that batch mode is still much slower than using the native batch mode of exiftool on the command line.

**Reply by sylikc** [reply]
> For me, I care about the ability of batch processing a ton of files at once. The primary reason I use the original library is that it uses the native batch mode of exiftool in Perl (kind of opening a server and keep sending it commands.) I’m not sure it is related to the things you’re modifying, but I can keep my own fork for that purpose.

oh, no, I'm not removing the native batch mode.  That's the best part!

The `-stay_open` flag is still there.  the `*_batch` functionality I'm talking about is the `get_metadata_batch()` and `get_metadata()` helper methods that smarnach had added to the ExifTool class for people to use easier.  As he said in the other thread, he really designed the code for illustration purposes, not really considering all the possible ways to break it lol.

I'm good at breaking things :).  Just filed an issue with CPython earlier today because I can't figure out why [ExifTool class hangs the interpreter on `__del__` in Windows](https://bugs.python.org/issue43784) (it appears to "potentially" be a bug in CPython 3.8-3.10, but the debugging part is beyond me)

> P.S. By my own test, that batch mode is still much slower than using the native batch mode of exiftool on the command line.

how are you using the batch mode on the command line.  **Can you write a simple benchmark for me to test against?**  I actually noticed how smarnach optimized memory and resource management lately as I've been trying to tweak the code.  I might be able to look into optimizing it if it's much slower.  I wonder why it would be...

### Example of the _batch stuff I'm talking about
Edit: If you want to see what I mean by removing the `_batch` function and combining it into a single function you can see the `def get_tags(self, in_tags, in_files)` in the [latest version on the v0.5 branch](https://github.com/sylikc/pyexiftool/blob/v0.5.x-py3-refactor/exiftool/helper.py#L177)

This combines the functionality of the original [def get_tags_batch(self, tags, filenames)](https://github.com/smarnach/pyexiftool/blob/master/exiftool.py#L270) and the original [def get_tags(self, tags, filename)](https://github.com/smarnach/pyexiftool/blob/master/exiftool.py#L293) into one very usable (and error-checked to death) function with a consistent return value

**Reply by ickc** [reply]
I can copy the exact command I used in the past if you need, but the command line batch mode I mentioned means calling exiftool on a glob pattern and with a certain arg that I don’t remember right now, ask it to write the exif data as a JSON sidecar. Since I read a lot of exif (100,000 or more) I found that caching them as a JSON sidecar first is much quicker. Eg you can call exiftool directly to write those files first, and later in your script reading from those JSON, is much faster than calling it through pyexiftool batch mode directly, even when doing exactly the same thing. (That’s also why I submitted a patch that fix a bug preventing pyexiftool to writes to JSON sidecar.)

however, you may not need to worry about this, as it might be an inherent speed limitation in running exiftool in “server mode”. Often, calling across programming interface adds latency that is especially noticeable in tight loop (Ie the latency is amplified by the no of times you call it, and is a significant faction of the total run time.)

**Comment by sylikc** [comment]
Alright, I have a draft PR created https://github.com/sylikc/pyexiftool/pull/13 to start the discussion with direct looks at the code.

Could you take some time to see if PyExifTool is moving in a direction you'd be happy with? 😊

The base `ExifTool` class is pretty much finalized, except for the TODO comments, and the addition of a logging facility coming.

The `ExifToolHelper` class has lots of work still... that's not even close to finalized.

**Reply by csparker247** [reply]
Great! I've got a big deadline this week, but will put it on my list for review next week.
